Overriding entire pipeline fails

### Summary 
I am trying to run a custom pipeline that worked previously with the `yaml` below.  Now I get an error as shown below.

**NOTE:**  I noticed if I remove the `buildPack: none` at the top of the yaml file, then it does not error out.

### Steps to reproduce the behavior
Simply add the `yaml` to your `jenkins-x.yml` file and do a pull request.

### Expected behavior
The pipeline runs and the logs shows the custom steps. Using the following definition

```yaml
buildPack: none
pipelineConfig:
  pipelines:
    release:
      pipeline:
        agent:
          image: sharepointoscar/node:8
        stages:
          - name: Oscar Build Stage
            steps:
              - name: Get Node Version
                command: node
                args:
                  - --version
              - name: NPM Installar Por Favor
                command: npm
                args: 
                  - install               
            options:
              containerOptions:
                resources:
                  limits:
                    cpu: 0.4
                    memory: 256Mi
                  requests:
                    cpu: 0.2
                    memory: 128Mi
          - name: Oscar Test Stage
            steps:
              - name: Run MochaJS Tests
                command: npm
                args:
                  - test
            options:
              containerOptions:
                resources:
                  limits:
                    cpu: 0.4
                    memory: 256Mi
                  requests:
                    cpu: 0.2
                    memory: 128Mi
```

### Actual behavior
I get the error listed below.

```
Showing logs for build jenkins-oscar/sp-webpart/pr-1 #14 serverless-jenkins stage app-extension and container step-create-effective-pipeline
{"level":"info","msg":"Deleting and cloning the Jenkins X versions repo","time":"2019-08-22T23:26:27Z"}
{"level":"info","msg":"Cloning the Jenkins X versions repo https://github.com/jenkins-x/jenkins-x-versions.git with ref refs/heads/master to /builder/home/.jx/jenkins-x-versions","time":"2019-08-22T23:26:27Z"}
{"level":"info","msg":"Effective pipeline written to jenkins-x-serverless-jenkins-effective.yml","time":"2019-08-22T23:26:29Z"}

Showing logs for build jenkins-oscar/sp-webpart/pr-1 #14 serverless-jenkins stage app-extension and container step-create-tekton-crds
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x11dcb4a]

goroutine 1 [running]:
github.com/jenkins-x/jx/pkg/tekton/syntax.(*ParsedPipeline).GenerateCRDs(0x0, 0xc00050b1a0, 0x25, 0x7ffde6032651, 0x2, 0xc000dc2440, 0x1f, 0xc000645ddc, 0x2, 0xc0005461e0, ...)
	/workspace/source/pkg/tekton/syntax/pipeline.go:1523 +0x3a
github.com/jenkins-x/jx/pkg/cmd/step/create.(*StepCreateTaskOptions).generateTektonCRDs(0xc0003222c0, 0xc0006c0630, 0xc000645ddc, 0x2, 0xc00050b1a0, 0x25, 0xc000dc2440, 0x1f, 0x0, 0x53267c0, ...)
	/workspace/source/pkg/cmd/step/create/step_create_task.go:543 +0x18f
github.com/jenkins-x/jx/pkg/cmd/step/create.(*StepCreateTaskOptions).Run(0xc0003222c0, 0x0, 0x0)
	/workspace/source/pkg/cmd/step/create/step_create_task.go:323 +0x66a
github.com/jenkins-x/jx/pkg/cmd/step/create.NewCmdStepCreateTaskAndOption.func1(0xc000624500, 0xc0000bc380, 0x8, 0x34)
	/workspace/source/pkg/cmd/step/create/step_create_task.go:146 +0x6f
github.com/spf13/cobra.(*Command).execute(0xc000624500, 0xc0000bc000, 0x34, 0x38, 0xc000624500, 0xc0000bc000)
	/workspace/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:766 +0x2ae
github.com/spf13/cobra.(*Command).ExecuteC(0xc0002ae280, 0xc00034c550, 0x530c2c0, 0xc00000e010)
	/workspace/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852 +0x2ec
github.com/spf13/cobra.(*Command).Execute(...)
	/workspace/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
github.com/jenkins-x/jx/cmd/jx/app.Run(0x0, 0x0, 0x0, 0xc0009eff88, 0xc0000a0058)
	/workspace/source/cmd/jx/app/jx.go:19 +0x147
main.main()
	/workspace/source/cmd/jx/jx.go:11 +0x32

Pipeline failed on stage 'app-extension' : container 'step-create-tekton-crds'. The execution of the pipeline has stopped.
```

### Jx version

The output of `jx version` is:

```
NAME               VERSION
jx                 2.0.598
jenkins x platform 2.0.951
Kubernetes cluster v1.12.8-gke.10
kubectl            v1.15.3
helm client        Client: v2.13.1+g618447c
git                2.21.0
Operating System   Mac OS X 10.14.3 build 18D42
```

### Jenkins type

<!--
Select which installation type are you using. 
-->
- [ x] Serverless Jenkins X Pipelines (Tekton + Prow)
- [ ] Classic Jenkins

### Kubernetes cluster
What kind of Kubernetes cluster are you using & how did you create it?
Using GKE, used the following command to create cluster:

```bash
jx create cluster -n mycluster --tekton --no-tiller --default-admin-password=mypassword
```


### Operating system / Environment

In which environment are you running the jx CLI?
OS X Mojave 10.14.3

--- a/pkg/cmd/step/create/step_create_task_test.go
+++ b/pkg/cmd/step/create/step_create_task_test.go
@@ -344,7 +344,7 @@ func TestGenerateTektonCRDs(t *testing.T) {
 			organization:          "abayer",
 			branch:                "really-long",
 			kind:                  "release",
-			effectiveProjectError: fmt.Errorf("%s", "no pipeline defined for kind release"),
+			effectiveProjectError: errors.New("no pipeline defined for kind release"),
 		},
 	}
 

--- a/pkg/cmd/step/create/step_create_task_test.go
+++ b/pkg/cmd/step/create/step_create_task_test.go
@@ -1,6 +1,7 @@
 package create
 
 import (
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"os"
@@ -42,14 +43,15 @@ import (
 )
 
 type testCase struct {
-	name           string
-	language       string
-	repoName       string
-	organization   string
-	branch         string
-	kind           string
-	expectingError bool
-	useKaniko      bool
+	name                  string
+	language              string
+	repoName              string
+	organization          string
+	branch                string
+	kind                  string
+	generateError         error
+	effectiveProjectError error
+	useKaniko             bool
 }
 
 func TestGenerateTektonCRDs(t *testing.T) {
@@ -104,13 +106,13 @@ func TestGenerateTektonCRDs(t *testing.T) {
 			kind:         "release",
 		},
 		{
-			name:           "no_pipeline_config",
-			language:       "none",
-			repoName:       "anything",
-			organization:   "anything",
-			branch:         "anything",
-			kind:           "release",
-			expectingError: true,
+			name:                  "no_pipeline_config",
+			language:              "none",
+			repoName:              "anything",
+			organization:          "anything",
+			branch:                "anything",
+			kind:                  "release",
+			effectiveProjectError: errors.New("effective pipeline creation failed: failed to find PipelineConfig in file test_data/step_create_task/no_pipeline_config/jenkins-x.yml"),
 		},
 		{
 			name:         "per_step_container_build_pack",
@@ -228,13 +230,13 @@ func TestGenerateTektonCRDs(t *testing.T) {
 			kind:         "release",
 		},
 		{
-			name:           "remove-pipeline",
-			language:       "none",
-			repoName:       "anything",
-			organization:   "anything",
-			branch:         "anything",
-			kind:           "pullRequest",
-			expectingError: true,
+			name:                  "remove-pipeline",
+			language:              "none",
+			repoName:              "anything",
+			organization:          "anything",
+			branch:                "anything",
+			kind:                  "pullrequest",
+			effectiveProjectError: errors.New("no pipeline defined for kind pullrequest"),
 		},
 		{
 			name:         "remove-stage-from-jenkins-x-yml",
@@ -245,13 +247,13 @@ func TestGenerateTektonCRDs(t *testing.T) {
 			kind:         "release",
 		},
 		{
-			name:           "remove-pipeline-from-jenkins-x-yml",
-			language:       "none",
-			repoName:       "anything",
-			organization:   "anything",
-			branch:         "anything",
-			kind:           "pullRequest",
-			expectingError: true,
+			name:                  "remove-pipeline-from-jenkins-x-yml",
+			language:              "none",
+			repoName:              "anything",
+			organization:          "anything",
+			branch:                "anything",
+			kind:                  "pullrequest",
+			effectiveProjectError: errors.New("no pipeline defined for kind pullrequest"),
 		},
 		{
 			name:         "replace-stage-steps",
@@ -335,6 +337,15 @@ func TestGenerateTektonCRDs(t *testing.T) {
 			branch:       "really-long",
 			kind:         "release",
 		},
+		{
+			name:                  "no_pipeline_for_kind",
+			language:              "none",
+			repoName:              "js-test-repo",
+			organization:          "abayer",
+			branch:                "really-long",
+			kind:                  "release",
+			effectiveProjectError: fmt.Errorf("%s", "no pipeline defined for kind release"),
+		},
 	}
 
 	k8sObjects := []runtime.Object{
@@ -408,33 +419,43 @@ func TestGenerateTektonCRDs(t *testing.T) {
 			assert.NoError(t, err)
 
 			effectiveProjectConfig, err := createTask.createEffectiveProjectConfig(packsDir, projectConfig, projectConfigFile, resolver, ns)
-			if err != nil && strings.Contains(err.Error(), "validation failed for Pipeline") {
-				t.Fatalf("Validation failure for effective pipeline: %s", err)
-			}
-			if effectiveProjectConfig != nil {
-				err = createTask.setBuildVersion(effectiveProjectConfig.PipelineConfig)
-				assert.NoError(t, err)
-			}
-
-			resourceName := tekton.PipelineResourceNameFromGitInfo(createTask.GitInfo, createTask.Branch, createTask.Context, tekton.BuildPipeline, nil, "")
-			pipelineName := tekton.PipelineResourceNameFromGitInfo(createTask.GitInfo, createTask.Branch, createTask.Context, tekton.BuildPipeline, tektonClient, ns)
-			crds, err := createTask.generateTektonCRDs(effectiveProjectConfig, ns, pipelineName, resourceName)
-			if tt.expectingError {
+			if tt.effectiveProjectError != nil {
 				if err == nil {
-					t.Fatalf("Expected an error generating CRDs")
+					t.Fatalf("Expected an error %s generating effective project config, did not see it", tt.effectiveProjectError)
 				}
+				assert.Equal(t, tt.effectiveProjectError.Error(), err.Error(), "Expected error %s but received error %s", tt.effectiveProjectError, err)
+			} else if err != nil {
+				if strings.Contains(err.Error(), "validation failed for Pipeline") {
+					t.Fatalf("Validation failure for effective pipeline: %s", err)
+				}
+				t.Fatalf("Unexpected error generating effective pipeline: %s", err)
 			} else {
-				if err != nil {
-					t.Fatalf("Error generating CRDs: %s", err)
-				}
-
-				// to update the golden files 'make test1-pkg PKG=./pkg/cmd/step/create TEST=TestGenerateTektonCRDs UPDATE_GOLDEN=1' - use with care!
-				if os.Getenv("UPDATE_GOLDEN") != "" {
-					err = crds.WriteToDisk(caseDir, nil)
+				if effectiveProjectConfig != nil {
+					err = createTask.setBuildVersion(effectiveProjectConfig.PipelineConfig)
 					assert.NoError(t, err)
 				}
 
-				assertTektonCRDs(t, tt, crds, caseDir, createTask)
+				resourceName := tekton.PipelineResourceNameFromGitInfo(createTask.GitInfo, createTask.Branch, createTask.Context, tekton.BuildPipeline, nil, "")
+				pipelineName := tekton.PipelineResourceNameFromGitInfo(createTask.GitInfo, createTask.Branch, createTask.Context, tekton.BuildPipeline, tektonClient, ns)
+				crds, err := createTask.generateTektonCRDs(effectiveProjectConfig, ns, pipelineName, resourceName)
+				if tt.generateError != nil {
+					if err == nil {
+						t.Fatalf("Expected an error %s generating CRDs, did not see it", tt.generateError)
+					}
+					assert.Equal(t, tt.generateError, err, "Expected error %s but received error %s", tt.generateError, err)
+				} else {
+					if err != nil {
+						t.Fatalf("Unexpected error generating CRDs: %s", err)
+					}
+
+					// to update the golden files 'make test1-pkg PKG=./pkg/cmd/step/create TEST=TestGenerateTektonCRDs UPDATE_GOLDEN=1' - use with care!
+					if os.Getenv("UPDATE_GOLDEN") != "" {
+						err = crds.WriteToDisk(caseDir, nil)
+						assert.NoError(t, err)
+					}
+
+					assertTektonCRDs(t, tt, crds, caseDir, createTask)
+				}
 			}
 		})
 	}

--- a/pkg/cmd/step/create/test_data/step_create_task/no_pipeline_for_kind/jenkins-x.yml
+++ b/pkg/cmd/step/create/test_data/step_create_task/no_pipeline_for_kind/jenkins-x.yml
@@ -1,0 +1,42 @@
+buildPack: none
+pipelineConfig:
+  env:
+    - name: FRUIT
+      value: BANANA
+  pipelines:
+    pullRequest:
+      pipeline:
+        env:
+          - name: GIT_AUTHOR_NAME
+            value: somebodyelse
+        options:
+          containerOptions:
+            resources:
+              requests:
+                cpu: 0.1
+                memory: 64Mi
+        agent:
+          image: nodejs
+        stages:
+          - name: Build-a-really-long-stage-name-please-but-not-too-long-thanks
+            steps:
+              - command: echo
+                args:
+                  - hello world
+              - command: ls
+                args:
+                  - -la
+                env:
+                  - name: ANOTHER_VAR
+                    value: Another value
+          - name: Second
+            steps:
+              - command: echo
+                args:
+                  - hi ${FRUIT}
+            options:
+              containerOptions:
+                resources:
+                  limits:
+                    cpu: 0.4
+                    memory: 256Mi

--- a/pkg/cmd/step/create/test_data/step_create_task/remove-pipeline-from-jenkins-x-yml/jenkins-x.yml
+++ b/pkg/cmd/step/create/test_data/step_create_task/remove-pipeline-from-jenkins-x-yml/jenkins-x.yml
@@ -1,3 +1,4 @@
+buildPack: none
 pipelineConfig:
   env:
     - name: FRUIT

--- a/pkg/cmd/step/syntax/step_syntax_effective.go
+++ b/pkg/cmd/step/syntax/step_syntax_effective.go
@@ -388,6 +388,11 @@ func (o *StepSyntaxEffectiveOptions) createPipelineForKind(kind string, lifecycl
 		if projectConfig.BuildPack == "" || projectConfig.BuildPack == "none" {
 			for _, override := range pipelines.Overrides {
 				if override.MatchesPipeline(kind) {
+					// If no step/steps, other overrides, or stage is specified, just remove the whole pipeline.
+					// TODO: This is probably pointless functionality.
+					if override.Step == nil && len(override.Steps) == 0 && !override.HasNonStepOverrides() && override.Stage == "" {
+						return nil, nil
+					}
 					parsed = syntax.ApplyStepOverridesToPipeline(parsed, override)
 				}
 			}

--- a/pkg/config/project_config.go
+++ b/pkg/config/project_config.go
@@ -159,11 +159,11 @@ func (c *ProjectConfig) GetPipeline(kind string) (*syntax.ParsedPipeline, error)
 			parsed = c.PipelineConfig.Pipelines.Feature.Pipeline
 		}
 	default:
-		return nil, fmt.Errorf("unknown pipeline kind %s", kind)
+		return nil, errors.Errorf("unknown pipeline kind %s", kind)
 	}
 
 	if parsed == nil {
-		return nil, fmt.Errorf("no pipeline defined for kind %s", kind)
+		return nil, errors.Errorf("no pipeline defined for kind %s", kind)
 	}
 	return parsed, nil
 }

--- a/pkg/config/project_config.go
+++ b/pkg/config/project_config.go
@@ -161,5 +161,9 @@ func (c *ProjectConfig) GetPipeline(kind string) (*syntax.ParsedPipeline, error)
 	default:
 		return nil, fmt.Errorf("unknown pipeline kind %s", kind)
 	}
+
+	if parsed == nil {
+		return nil, fmt.Errorf("no pipeline defined for kind %s", kind)
+	}
 	return parsed, nil
 }

--- a/pkg/config/project_config_test.go
+++ b/pkg/config/project_config_test.go
@@ -93,12 +93,12 @@ func TestGetPipeline(t *testing.T) {
 	assert.NotNil(t, releasePipeline)
 
 	pullRequestPipeline, err := testProjectConfigMaven.GetPipeline(jenkinsfile.PipelineKindPullRequest)
-	assert.NotNil(t, err)
+	assert.Error(t, err)
 	assert.Equal(t, err.Error(), "no pipeline defined for kind pullrequest")
 	assert.Nil(t, pullRequestPipeline)
 
 	featurePipeline, err := testProjectConfigMaven.GetPipeline(jenkinsfile.PipelineKindFeature)
-	assert.NotNil(t, err)
+	assert.Error(t, err)
 	assert.Equal(t, err.Error(), "no pipeline defined for kind feature")
 	assert.Nil(t, featurePipeline)
 }

--- a/pkg/config/project_config_test.go
+++ b/pkg/config/project_config_test.go
@@ -93,10 +93,12 @@ func TestGetPipeline(t *testing.T) {
 	assert.NotNil(t, releasePipeline)
 
 	pullRequestPipeline, err := testProjectConfigMaven.GetPipeline(jenkinsfile.PipelineKindPullRequest)
-	assert.NoError(t, err)
+	assert.NotNil(t, err)
+	assert.Equal(t, err.Error(), "no pipeline defined for kind pullrequest")
 	assert.Nil(t, pullRequestPipeline)
 
 	featurePipeline, err := testProjectConfigMaven.GetPipeline(jenkinsfile.PipelineKindFeature)
-	assert.NoError(t, err)
+	assert.NotNil(t, err)
+	assert.Equal(t, err.Error(), "no pipeline defined for kind feature")
 	assert.Nil(t, featurePipeline)
 }

--- a/pkg/tekton/syntax/pipeline.go
+++ b/pkg/tekton/syntax/pipeline.go
@@ -1884,7 +1884,7 @@ func (p *PipelineOverride) AsStepsSlice() []*Step {
 
 // MatchesPipeline returns true if the pipeline name is specified in the override or no pipeline is specified at all in the override
 func (p *PipelineOverride) MatchesPipeline(name string) bool {
-	if p.Pipeline == "" || p.Pipeline == name {
+	if p.Pipeline == "" || strings.EqualFold(p.Pipeline, name) {
 		return true
 	}
 	return false

